python/taichi/ui/staging_buffer.py

```diff
@@ -93,25 +94,81 @@ def copy_colors_to_vbo(vbo, colors):
 
 @ti.kernel
 def copy_image_f32_to_rgba8(src: ti.template(), dst: ti.template(),
-                            num_components: ti.template()):
-    for i, j in src:
+                            num_components: ti.template(),
+                            gray_scale: ti.template()):
+    for i, j in ti.ndrange(src.shape[0], src.shape[1]):
         px = ti.Vector([0, 0, 0, 0xff], dt=u32)
-        for k in ti.static(range(num_components)):
-            c = src[i, j][k]
+        if ti.static(gray_scale):
+            c = 0.0
+            c = src[i, j]
             c = max(0.0, min(1.0, c))
             c = c * 255
-            px[k] = ti.cast(c, u32)
+            px[0] = px[1] = px[2] = ti.cast(c, u32)
+        else:
+            for k in ti.static(range(num_components)):
+                c = 0.0
+                if ti.static(len(src.shape) == 3):
```

Comment:
Maybe split this code into 2 functions to handle different field type is good to read?

Reply:
Yeah, I had considered about this when writing the code. On a second thought, the only difference between operating `ti.field(shape=(width, height, 3))` and `ti.Vector.field(3, shape=(width, height))` is the `k`'s position. In order to simplify the code, I made the decision to merge the 2 functions into one and use static branching to eliminate runtime overhead. Perhaps I could write some comments to make this more readable?

Reply:
That's right, thanks!

Reply:
Yes, it's good to add some notes to remind others. 

And still we'd better handle different type of data in different functions instead of taking a trick.